src/core/comm/coll.cc

```diff
@@ -0,0 +1,389 @@
+                  CollDataType type,
+                  CollComm global_comm)
+{
+  log_coll.print("Alltoallv: global_rank %d, mpi_rank %d, unique_id %d, comm_size %d",
+                 global_comm->global_rank,
+                 global_comm->mpi_rank,
+                 global_comm->unique_id,
+                 global_comm->global_comm_size);
+#ifdef LEGATE_USE_GASNET
+  return alltoallvMPI(
+    sendbuf, sendcounts, sdispls, recvbuf, recvcounts, rdispls, type, global_comm);
+#else
+  return alltoallvLocal(
+    sendbuf, sendcounts, sdispls, recvbuf, recvcounts, rdispls, type, global_comm);
+#endif
+}
+
+int collAlltoall(
+  const void* sendbuf, void* recvbuf, int count, CollDataType type, CollComm global_comm)
+{
+  log_coll.print("Alltoall: global_rank %d, mpi_rank %d, unique_id %d, comm_size %d",
+                 global_comm->global_rank,
+                 global_comm->mpi_rank,
+                 global_comm->unique_id,
+                 global_comm->global_comm_size);
+#ifdef LEGATE_USE_GASNET
+  return alltoallMPI(sendbuf, recvbuf, count, type, global_comm);
+#else
+  return alltoallLocal(sendbuf, recvbuf, count, type, global_comm);
+#endif
+}
+
+int collAllgather(
+  const void* sendbuf, void* recvbuf, int count, CollDataType type, CollComm global_comm)
+{
+  log_coll.print("Allgather: global_rank %d, mpi_rank %d, unique_id %d, comm_size %d",
+                 global_comm->global_rank,
+                 global_comm->mpi_rank,
+                 global_comm->unique_id,
+                 global_comm->global_comm_size);
+#ifdef LEGATE_USE_GASNET
+  return allgatherMPI(sendbuf, recvbuf, count, type, global_comm);
+#else
+  return allgatherLocal(sendbuf, recvbuf, count, type, global_comm);
+#endif
+}
+
+// called from main thread
+int collInit(int argc, char* argv[])
+{
+  current_unique_id    = 0;
+  const char* nb_comms = getenv("LEGATE_MAX_COMMS");
+  if (nb_comms != nullptr) { MAX_NB_COMMS = atoi(nb_comms); }
+  assert(MAX_NB_COMMS > 0);
+#ifdef LEGATE_USE_GASNET
+  int provided, init_flag = 0;
+  CHECK_MPI(MPI_Initialized(&init_flag));
+  if (!init_flag) {
+    CHECK_MPI(MPI_Init_thread(&argc, &argv, MPI_THREAD_MULTIPLE, &provided));
+  } else {
+    log_coll.print(
+      "Warning: MPI has been initialized by others, make sure MPI is initialized with "
+      "MPI_THREAD_MULTIPLE");
+  }
+  mpi_comms.resize(MAX_NB_COMMS, MPI_COMM_NULL);
+  for (int i = 0; i < MAX_NB_COMMS; i++) { CHECK_MPI(MPI_Comm_dup(MPI_COMM_WORLD, &mpi_comms[i])); }
+#else
+  thread_comms.resize(MAX_NB_COMMS);
+  for (int i = 0; i < MAX_NB_COMMS; i++) {
+    thread_comms[i].ready_flag = false;
+    thread_comms[i].buffers    = nullptr;
+    thread_comms[i].displs     = nullptr;
+  }
+#endif
+  coll_inited = true;
+  return CollSuccess;
+}
+
+int collFinalize(void)
+{
+  assert(coll_inited == true);
+  coll_inited = false;
+#ifdef LEGATE_USE_GASNET
+  for (int i = 0; i < MAX_NB_COMMS; i++) { CHECK_MPI(MPI_Comm_free(&mpi_comms[i])); }
```

Comment:
So this is a step that I think should be performed normally by the finalize task of this communicator. I don't think it's a good design to defer all communicator destructions to the end of the program execution.

Reply:
The user-level `CollComm` is destroyed by the `collCommDestroy` call in the finalize task, but not the `MPI_Comm`. We can reuse the `MPI_Comm` for later created `CollComm`.

Reply:
Sure, but do we actually reuse the existing `MPI_Comm`? I don't think I see the code that does that.

Reply:
Actually I have the reuse code in another branch
https://github.com/eddy16112/legate.core/blob/coll-pool/src/core/comm/coll.cc#L278
Since I realized legate won't delete any communicators until the finalization of the library, so it is necessary to have the reuse code, which can increase the complexity.

Reply:
Legate currently doesn't destroy any communicator during execution, but that can change in the future. And Legate's communicator manager is already pooling the communicators, so I don't think individual communicators should roll their own pooling logic like this. I'd rather have the `MPI_Comm` destroyed when the wrapping `CollComm` is destroyed.

Reply:
I think we can postpone this problem together with the vector of `MPI_Comm` because they are actually the same one.